### ISSUE TITLE
normalize_hw / std_hw / var_hw: the variance composite returns an exact 0 below |x-mean| ~1e-19 and saturates above ~1.8e19

### ISSUE THREAD

**Issue by EazyHood** [tenstorrent/tt-metal#52614]
### Component / Area

ttnn eltwise unary composites - ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp (var_hw, std_hw, normalize_hw, normalize_global)

### Issue Type (optional)

Bad Outputs

### Observed

`ttnn.std_hw` and `ttnn.var_hw` return exactly zero for a spread of ordinary normal values, and
return the same fixed constant for every spread above about `1.3e19`. `ttnn.normalize_hw` divides
by that standard deviation, so it is non-finite at the low end and grows linearly at the high end,
where the correct answer is `±1.0` at every magnitude.

Input used throughout: a `[1, 1, 32, 32]` `float32` tile whose columns alternate `+v` and `-v`.
The mean is exactly `0` and the population standard deviation is exactly `v`, so `std_hw` must
return `v` and `normalize_hw` must return the `±1` pattern for every `v`.

### Low end - an exact zero, not a tolerance question

| deviation `v` | `std_hw` | exact | `normalize_hw` |
|---|---|---|---|
| 1e-25 | **0** | 1e-25 | not finite |
| 1e-20 | **0** | 1e-20 | not finite |
| 1.0842e-19 | **0** | 1.0842e-19 | not finite |
| 1e18 | 9.99759e17 | 1e18 | 1.00004 (ok) |

`1e-20` is an ordinary normal `float32`, nowhere near the subnormal boundary. The square is what
leaves the range: `(1.0842e-19)^2 = 1.1754e-38` is the smallest normal `float32`, so anything
below that flushes to zero, the mean of zeros is zero, and the square root of zero is zero.
`1.0842e-19` is `sqrt(FLT_MIN)` - a property of the intermediate, not of the answer.

### High end - a fixed ceiling, not an overflow

| deviation `v` | `std_hw` | exact | `normalize_hw` |
|---|---|---|---|
| 1.8e19 | 1.30438e19 | 1.8e19 | 1.376 |
| 1.9e19 | 1.30438e19 | 1.9e19 | 1.447 |
| 1e20 | 1.30438e19 | 1e20 | **7.646** |
| 1e25 | 1.30438e19 | 1e25 | **764626** |

`std_hw` returns the same value for every deviation from about `1.3e19` upwards. That constant is
`sqrt(2^127) = 1.3043818e19`: the squared term tops out at `2^127`, the largest power of two in
`float32`, and the square root reports the ceiling. So `normalize_hw` up there is not merely
imprecise - it scales linearly with the input while the correct answer is fixed at `1.0`.

Worth being explicit about which claim is stronger. The exact zero at the low end follows from
`float32` alone. The high end is a saturation whose exact ceiling could differ on silicon; the
shape - a fixed ceiling, and therefore `normalize_hw` growing linearly - should not.

### Where it comes from

`ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp`, `_variance_impl`, forms
the square as a value of its own:

```cpp
Tensor sqr_y_minus_mean_y = ttnn::square(y_minus_mean_y, output_mem_config);
return ttnn::sum(sqr_y_minus_mean_y, dims, true, std::nullopt, std::nullopt, scale);
```

`std_hw` is `sqrt(var_hw)`, `normalize_hw` is `(y - mean) * reciprocal(std)`, and
`normalize_global` is built from `normalize_hw`, so all four inherit it. The final result - a
standard deviation of `1e-20` or `1e20` - is an ordinary `float32` in both cases. Only the
intermediate is out of range.

### Expected

For a tile whose elements are `±v` around a zero mean, `normalize_hw` should return the `±1`
pattern for every `v` representable as a normal number, in `float32` and `bfloat16` - not a
non-finite value below `1.0842e-19`, and not a value that grows with `v` above `1.3043818e19`.

`std_hw(k * x)` should equal `k * std_hw(x)` to the op's existing tolerance, for `k` spanning the
exponent range. Standard deviation is homogeneous of degree one, so this holds by construction; it
is the cheapest oracle available for these ops, and it is not asserted anywhere today.

`std_hw` and `var_hw` should track the exact value across that range instead of returning `0` or a
fixed constant, and `var_hw` and `normalize_global` should be covered too, since they share the
composition. Values in the currently working band should be unchanged.

One limit worth stating up front rather than leaving it to be found: `var_hw` cannot return
`var(1e-20) = 1e-40`, because that number is not a `float32`. The low end is fixable for `std_hw`
and `normalize_hw`, whose results are ordinary `float32` values; for `var_hw` only the high end is
- where the variance is representable and currently comes back as a saturated constant.

### 1. Steps (exact commands)

I do not have a Tenstorrent card, so these are the commands for someone who does; what I ran is
described under Hardware Details.

Save as `repro_normalize_hw.py` at the repo root:

```python
import torch
import ttnn


def tile(v):
    # [1, 1, 32, 32] with columns alternating +v and -v.
    # Mean is exactly 0; population standard deviation is exactly v.
    t = torch.full((1, 1, 32, 32), v, dtype=torch.float32)
    t[..., 1::2] = -v
    return t


device = ttnn.open_device(device_id=0)

print(f"{'deviation':>12} {'std_hw':>14} {'exact':>12} {'normalize_hw':>14}")
for v in [1e-25, 1e-20, 1.0842e-19, 1e18, 1.8e19, 1.9e19, 1e20, 1e25]:
    x = ttnn.from_torch(tile(v), dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
    std = ttnn.to_torch(ttnn.from_device(ttnn.std_hw(x)))[0, 0, 0, 0].item()
    nrm = ttnn.to_torch(ttnn.from_device(ttnn.normalize_hw(x)))[0, 0, 0, 0].item()
    print(f"{v:12.4e} {std:14.6e} {v:12.4e} {nrm:14.6g}")

ttnn.close_device(device)
```

Run it:

```bash
python repro_normalize_hw.py
```

A correct implementation prints `std_hw` equal to the `exact` column on every row, and
`normalize_hw` equal to `1.0` on every row. What comes out today is under Logs & Diagnostics.

Swapping `ttnn.float32` for `ttnn.bfloat16` gives the same two failures: the two types share an
exponent field, so they share both thresholds.

### 2. Input data / link or description

None. The repro builds its own tensors; there is no external file, dataset or model to link.

The construction is the reference on purpose. A `[1, 1, 32, 32]` tile whose columns alternate `+v`
and `-v` has a mean of exactly `0` and a population standard deviation of exactly `v`, by algebra,
at every magnitude. So `std_hw` must return `v` and `normalize_hw` must return the same `±1`
pattern for every `v`, with no tolerance argument needed.

`torch` is used only to build the input. It is not the reference here: `torch.std` in `float32`
forms the same square and overflows at the top of this range, so it cannot arbitrate the high end.
The analytic value can.

Deviations used: `1e-25`, `1e-20`, `1.0842e-19`, `1e18`, `1.8e19`, `1.9e19`, `1e20`, `1e25`. Each
one is a normal `float32`.

### 3. Frequency

Consistent - deterministic, every run.

Both thresholds are properties of `float32` arithmetic, not of scheduling or of a random draw:
below `sqrt(FLT_MIN) = 1.0842e-19` the squared term falls under the smallest normal and flushes to
zero, and above `sqrt(2^127) = 1.3043818e19` it reaches the ceiling. The repro seeds nothing and
draws nothing at random, so the same `v` gives the same result on every run.

### 1. Software Versions

| | |
|---|---|
| tt-metal | `main` at `68b985e7ed5164faf2ac95ba3a0530c8a1488482` (2026-08-04) |
| torch | 2.13.0+cpu |
| Python | 3.14 |
| ttsim | 1.9.7, `Arch.BLACKHOLE` |

The source lines quoted under Observed are from that SHA.

### 2. Hardware Details

No Tenstorrent card here. Saying so up front rather than letting a reviewer discover it.

Where the numbers come from:

- **The analysis is of the source** at the SHA above - `_variance_impl` in
  `unary_composite_op.cpp`, and the three ops built on it.
- **The full-range numbers come from reimplementing the composite's arithmetic in `float32`**:
  the same subtract, square, mean, square root, with flush-to-zero below the smallest normal and
  the `2^127` ceiling modelled. Over 122 deviations from `1e-30` to `1e30`, `std_hw` is wrong on
  46 of them. That reimplementation reproduces both measured boundaries exactly - `1.0842e-19` and
  `1.3043818e19` - which is why I use it for the range rather than only for the description.
- **The tile-level numbers under Observed and Logs were run in ttsim 1.9.7 with `Arch.BLACKHOLE`,
  `float32`.** That is the simulator, not silicon.
- `torch 2.13.0+cpu` builds the input tensor. It is not the reference: `torch.std` in `float32`
  forms the same square and overflows at the top of this range.

What that means for the two claims. The exact `0` below `1.0842e-19` follows from `float32`
itself, so it should reproduce on any card. The exact value of the high-end ceiling could differ
on silicon if the accumulator saturates somewhere other than `2^127`; the fixed-ceiling shape, and
`normalize_hw` growing linearly because of it, should not. A run of the repro on a Blackhole or
Wormhole card would settle the ceiling, and I would rather have that confirmation than assert it.

### Is this a regression?

No

### Regression Details

Not a regression that I can point to. `_variance_impl` computes the variance this way at the SHA
above; I did not bisect when it started, and I have no evidence of a commit that changed the
behaviour. What follows is context on adjacent work, not a blame.

- **#25100 and PR #41838** (merged 2026-04-14) reimplemented `ttnn.std` and `ttnn.var` with
  Welford's algorithm, for a failure of the same family - cancellation producing a negative
  variance or an infinite standard deviation. That work landed under `operations/reduction/generic`
  and `experimental/reduction`. None of it touches
  `ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp`, so the `_hw` and
  `_global` composites still carry the naive `E[(x-mean)^2]`.
- **Delegating to that Welford path would not fix this.** Welford addresses cancellation, not
  range: it still forms `delta * delta` and loses the same two ends.
- **PRs #7708 and #7748** (April 2024) raised comparison tolerances on these ops; the `std_hw`
  sweep still compares with `comp_allclose(atol=0.1)`. That is the precision axis. This is not
  that: an exact `0`, and a value that reaches `764626` where `1.0` is correct.
- Same shape as **#51861** (`hypot`, open) and **#52037** (`logaddexp`): a composition whose
  intermediate has a much narrower domain than its result. This one is in the reduction composite
  rather than an SFPU kernel, and in a different file from either.

### Logs & Diagnostics

Nothing is logged and no assertion fires. Both ends return a value, which is the problem.

What the repro prints, ttsim 1.9.7, `Arch.BLACKHOLE`, `float32`, `[1, 1, 32, 32]` tile of
alternating `+v` / `-v` columns (`exact` is `v` by construction):

| deviation | `std_hw` | exact | `normalize_hw` |
|---|---|---|---|
| 1e-25 | **0** | 1e-25 | not finite |
| 1e-20 | **0** | 1e-20 | not finite |
| 1.0842e-19 | **0** | 1.0842e-19 | not finite |
| 1e18 | 9.99759e17 | 1e18 | 1.00004 (ok) |
| 1.8e19 | **1.30438e19** | 1.8e19 | 1.376 |
| 1.9e19 | **1.30438e19** | 1.9e19 | 1.447 |
| 1e20 | **1.30438e19** | 1e20 | **7.646** |
| 1e25 | **1.30438e19** | 1e25 | **764626** |

Wider check, with the composite reimplemented in `float32` (flush-to-zero and the `2^127` ceiling
modelled), 122 deviations from `1e-30` to `1e30`: `std_hw` is wrong on **46 of 122**.

On a card the high-end column may show a different constant if the accumulator saturates
elsewhere; the low-end zeros should be identical.

### Priority

None

### Impact

Normalisation sits inside layer norm, attention scaling and feature preprocessing. Both failure
modes are the quiet kind: a zero standard deviation makes everything downstream non-finite, and
the fixed ceiling produces finite, plausible numbers that nothing will flag.

### Why the current tests do not catch it

- `tests/sweep_framework/sweeps/eltwise/unary/normalize_hw/normalize_hw.py` draws its input from
  `torch_random(low=-10, high=10)`.
- `tests/ttnn/python_api_testing/sweep_tests/test_configs/ci_sweep_tests_working/wormhole/ttnn_std_hw_test.yaml`
  and the matching `ttnn_normalize_hw_test.yaml` use `gen_rand` with `low: -10, high: 10`.
- `ttnn_normalize_global_test.yaml` and the `normalize_global` sweep use `-1` to `1`.
- The only direct test of `var_hw` is
  `tests/ttnn/docs_examples/test_eltwise_unary_examples.py::test_var_hw`, which logs the output
  and asserts nothing.
- `tests/ttnn/unit_tests/operations/reduce/test_reduction.py::test_std_var_hw_reduce_batch_crosses_tree_block`
  exercises `ttnn.std` / `ttnn.var` - the reduction path - not these composites.

So nothing samples `|x - mean|` outside roughly `[1e-19, 1.3e19]`, and no invariant is asserted
that would catch this independently of the values sampled. `std(k*x) == k*std(x)` would.

### On the fix

I have one ready, with a test that fails on `main` and passes with it, and will open it against
this issue. It factors the magnitude out of the deviations before squaring and restores the scale
after the square root, so only `k` has to be representable rather than `k^2`. It is not free: one
extra element-wise pass, one extra broadcast multiply and one extra reduction, roughly doubling
the variance step. The PR states that cost in its own section and offers a conditional version if
it is not acceptable.
